### PR TITLE
Support installing plugin SNAPSHOTs with SNASPHOT distribution

### PR DESCRIPTION
**Pull request by reta**
### Description
Support installing plugin SNAPSHOTs with SNASPHOT distribution

```
/opensearch-2.19.0-SNAPSHOT$ ./bin/opensearch-plugin install transport-reactor-netty4
-> Installing transport-reactor-netty4                                                                                                                                     
-> Downloading transport-reactor-netty4 from opensearch                                                                                                                    
[=================================================] 100%                                                                                                                   

```

### Related Issues
Part of https://github.com/opensearch-project/opensearch-build/issues/5096, addresses https://github.com/opensearch-project/OpenSearch/pull/16219

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by reta** [comment]
The signature part is not yet settled:

```
/opensearch-2.19.0-SNAPSHOT$ ./bin/opensearch-plugin install transport-reactor-netty4                                                                                                                                                                                                     
-> Installing transport-reactor-netty4                                                                                                                                     
-> Downloading transport-reactor-netty4 from opensearch                                                                                                                    
[=================================================] 100%                                                                                                                   
-> Failed installing transport-reactor-netty4                                                                                                                              
-> Rolling back transport-reactor-netty4                                                                                                                                   
-> Rolled back transport-reactor-netty4                                                                                                                                    
Exception in thread "main" java.io.IOException: Server returned HTTP response code: 403 for URL: https://artifacts.opensearch.org/snapshots/plugins/transport-reactor-netty4/2.19.0-SNAPSHOT/transport-reactor-netty4-2.19.0-SNAPSHOT.zip.sig
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream0(HttpURLConnection.java:2013)
        at java.base/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1614)
        at java.base/sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:223)
        at java.base/java.net.URL.openStream(URL.java:1325)                                                                                                                
        at org.opensearch.plugins.InstallPluginCommand.urlOpenStream(InstallPluginCommand.java:490)
        at org.opensearch.plugins.InstallPluginCommand.verifySignature(InstallPluginCommand.java:630)
        at org.opensearch.plugins.InstallPluginCommand.downloadAndValidate(InstallPluginCommand.java:608)
        at org.opensearch.plugins.InstallPluginCommand.download(InstallPluginCommand.java:316)
        at org.opensearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:271)
        at org.opensearch.plugins.InstallPluginCommand.execute(InstallPluginCommand.java:248)
        at org.opensearch.cli.EnvironmentAwareCommand.execute(EnvironmentAwareCommand.java:104)
        at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
        at org.opensearch.cli.MultiCommand.execute(MultiCommand.java:104)
        at org.opensearch.cli.Command.mainWithoutErrorHandling(Command.java:138)
        at org.opensearch.cli.Command.main(Command.java:101)
        at org.opensearch.plugins.PluginCli.main(PluginCli.java:60)

```

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 1]
```diff
@@ -341,38 +332,31 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir, boolean i
...
         final Terminal terminal,
-        final String stagingHash,
         final Version version,
         final boolean isSnapshot,
         final String pluginId,
         final String platform
     ) throws IOException, UserException {
         final String baseUrl;
-        if (isSnapshot && stagingHash == null) {
+        if (isSnapshot == false && version.isRelease() == false) {
```
Comment (rursprung) [inline distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:348]: why the `== false` & `== true`?! 😨

```suggestion
        if (!isSnapshot && !version.isRelease()) {
```

Comment (reta) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:348]: This is the convention we use across the whole codebase

Comment (rursprung) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:348]: oh, i hadn't noticed that yet! still appalled at it, though 🙈

Comment (andrross) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java]: @rursprung It's even called out in the [developer guide](https://github.com/opensearch-project/OpenSearch/blob/034bd2b6483c180b4a4439d62452cc50198c37fb/DEVELOPER_GUIDE.md?plain=1#L353)! (I'm equally appalled though for the record)

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 2]
```diff
@@ -341,38 +332,25 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir, boolean i
...
     private String getOpenSearchUrl(
         final Terminal terminal,
-        final String stagingHash,
         final Version version,
         final boolean isSnapshot,
         final String pluginId,
         final String platform
     ) throws IOException, UserException {
         final String baseUrl;
-        if (isSnapshot && stagingHash == null) {
```
[the remark this answers is not in the document]
Comment (cwperks) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:363]: so stagingHash is not applicable to OpenSearch? I suspected the same, but thank you for confirming.

[the remark this answers is not in the document]
Comment (reta) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:363]: correct, we don't publish artifacts with `stagingHash` or alike

**Review by dblock — APPROVED** [review]

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java [hunk 3]
```diff
@@ -341,38 +338,31 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir, boolean i
...
-        if (stagingHash != null) {
+        if (isSnapshot == true) {
             baseUrl = String.format(
                 Locale.ROOT,
-                "https://artifacts.opensearch.org/snapshots/plugins/%s/%s-%s",
+                "https://artifacts.opensearch.org/snapshots/plugins/%s/%s",
                 pluginId,
-                version,
-                stagingHash
+                version + "-SNAPSHOT"
```
[the remark this answers is not in the document]
Comment (saratvemulapalli) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:353]: Couldn't we use the same as release distributions  `Build.CURRENT.getQualifiedVersion()` ? 
Sounds like the only diff is the artifacts URL, could we consolidate the if-else ?

**Comment by reta** [comment]
Success! 
```
$ ./bin/opensearch-plugin install transport-nio
-> Installing transport-nio
-> Downloading transport-nio from opensearch
[=================================================] 100%   
-> Installed transport-nio with folder name transport-nio
```

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (reta) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:353]: Thanks @saratvemulapalli 

> This is tricky one: OpenSearch version has no SNAPSHOT suffix, but Build.CURRENT does

We don't pass build version here, only the `version` (which is current) + `isSnapshot`, I think it is fine (vs adding additional arguments)

 > Sounds like the only diff is the artifacts URL, could we consolidate the if-else ?

This if-else is only about  artifacts URL (`baseUrl`)

**Comment by reta** [comment]
```
$ /home/opensearch-3.0.0-SNAPSHOT$ ./bin/opensearch-plugin install transport-nio
-> Installing transport-nio
-> Downloading transport-nio from opensearch
[=================================================] 100%   
-> Installed transport-nio with folder name transport-nio

```

```
$ ll plugins/transport-nio/
total 3.3M
-rw-r--r-- 1 user group  12K Nov 12 14:15 LICENSE.txt
-rw-r--r-- 1 user group 329K Nov 12 14:15 netty-buffer-4.1.114.Final.jar
-rw-r--r-- 1 user group 344K Nov 12 14:15 netty-codec-4.1.114.Final.jar
-rw-r--r-- 1 user group 653K Nov 12 14:15 netty-codec-http-4.1.114.Final.jar
-rw-r--r-- 1 user group 679K Nov 12 14:15 netty-common-4.1.114.Final.jar
-rw-r--r-- 1 user group 564K Nov 12 14:15 netty-handler-4.1.114.Final.jar
-rw-r--r-- 1 user group  37K Nov 12 14:15 netty-resolver-4.1.114.Final.jar
-rw-r--r-- 1 user group 510K Nov 12 14:15 netty-transport-4.1.114.Final.jar
-rw-r--r-- 1 user group  44K Nov 12 14:15 netty-transport-native-unix-common-4.1.114.Final.jar
-rw-r--r-- 1 user group  16K Nov 12 14:15 NOTICE.txt
-rw-r--r-- 1 user group  59K Nov 12 14:15 opensearch-nio-3.0.0-SNAPSHOT.jar
-rw-r--r-- 1 user group 1.9K Nov 12 14:15 plugin-descriptor.properties
-rw-r--r-- 1 user group 1.7K Nov 12 14:15 plugin-security.policy
-rw-r--r-- 1 user group  66K Nov 12 14:15 transport-nio-client-3.0.0-SNAPSHOT.jar

```

**Comment by reta** [comment]
@saratvemulapalli LGTY? thanks!

File: distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (reta) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java:353]: Oh, I see what you mean, the `Build.CURRENT` could be used, indeed

[the remark this answers is not in the document]
Comment (saratvemulapalli) [reply distribution/tools/plugin-cli/src/main/java/org/opensearch/plugins/InstallPluginCommand.java]: Yup. Thank you.

**Review by saratvemulapalli — APPROVED** [review]

**Pull request merged**